crates/blind_pool/src/pool.rs

```diff
@@ -0,0 +1,563 @@
+
+    /// Drop policy that determines how the pool handles remaining items when dropped.
+    drop_policy: DropPolicy,
+}
+
+impl BlindPool {
+    /// Creates a new [`BlindPool`] with default configuration.
+    ///
+    /// This is equivalent to [`BlindPool::builder().build()`][BlindPool::builder].
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use blind_pool::BlindPool;
+    ///
+    /// let mut pool = BlindPool::new();
+    ///
+    /// let pooled = pool.insert(42_u64);
+    ///
+    /// // SAFETY: The pointer is valid and contains the value we just inserted.
+    /// let value = unsafe { pooled.ptr().read() };
+    /// assert_eq!(value, 42);
+    /// ```
+    #[must_use]
+    pub fn new() -> Self {
+        Self::builder().build()
+    }
+
+    /// Creates a builder for configuring and constructing a [`BlindPool`].
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use blind_pool::{BlindPool, DropPolicy};
+    ///
+    /// let pool = BlindPool::builder()
+    ///     .drop_policy(DropPolicy::MustNotDropItems)
+    ///     .build();
+    /// ```
+    pub fn builder() -> BlindPoolBuilder {
+        BlindPoolBuilder::new()
+    }
+
+    /// Creates a new [`BlindPool`] with the specified configuration.
+    ///
+    /// This method is used internally by the builder to construct the actual pool.
+    #[must_use]
+    pub(crate) fn new_inner(drop_policy: DropPolicy) -> Self {
+        Self {
+            pools: HashMap::with_hasher(FixedState::default()),
+            drop_policy,
+        }
+    }
+
+    /// Inserts a value into the pool and returns a handle to access it.
+    ///
+    /// The pool automatically determines the memory layout of the type and uses or creates
+    /// an appropriate internal pool to store the value.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use blind_pool::BlindPool;
+    ///
+    /// let mut pool = BlindPool::new();
+    ///
+    /// // Insert different types into the same pool.
+    /// let pooled_int = pool.insert(42_i32);
+    /// let pooled_float = pool.insert(2.5_f64);
+    /// let pooled_string = pool.insert("hello".to_string());
+    ///
+    /// // All values are stored in the same BlindPool.
+    /// assert_eq!(pool.len(), 3);
+    /// ```
+    pub fn insert<T>(&mut self, value: T) -> Pooled<T> {
+        let layout = Layout::new::<T>();
+
+        let internal_pool = self.pools.entry(layout).or_insert_with(|| {
+            OpaquePool::builder()
+                .layout_of::<T>()
+                .drop_policy(self.drop_policy)
+                .build()
+        });
+
+        // SAFETY: T matches the layout used to create the internal pool.
+        let pooled = unsafe { internal_pool.insert(value) };
+
+        Pooled { layout, pooled }
+    }
+
+    /// Removes a value from the pool and drops it.
+    ///
+    /// The [`Pooled<T>`] handle is consumed and the memory is returned to the pool.
+    /// The value is automatically dropped.
```

Comment:
The word "automatically" here is ambiguous - what makes it automatic? We called `remove()` to make it happen, so unclear what distinction it is making between "automatic" and "non-automatic" - better might be to omit this word.